# generic interactive plotter #
Generic interactive plotter using [Rshiny](https://shiny.rstudio.com/). The app can read in a file, and plot any X against any Y (as a scatterplot). The graph can be printed as a PDF for easy sharing.

![shinyplot_demo_screenshot.png](https://github.com/wrf/misc-analyses/blob/master/shinyplot/shinyplot_demo_screenshot.png)

---

---


# other shinyapps I have made #

# [blob plot contig viewer](https://github.com/wrf/lavaLampPlot/tree/master) #
Viewer for contig coverage vs GC content of genome/metagenome assemblies, for selecting contigs, or for printing the plot to PDF.

![blobplot viewer](https://github.com/wrf/lavaLampPlot/blob/164684cbd4c3dada50850a42708170edfadc470a/Rshiny/blobplot_screenshot_Emuelleri.png)

# [FAO Food balance explorer](https://github.com/wrf/foodbalance/tree/main) #
Explorer of FAO global food imports, production, and exports. Users can select and sort by country, food item, or trade direction.

![food balance app v2 screenshot](https://github.com/wrf/foodbalance/blob/e6ebe1d9ef0e285782e100b89ce56100ba4f592f/images/food_balance_app_v2_screenshot.png)

# [SRA global metagenome map](https://github.com/wrf/taxonomy_database/tree/master) #
Map of all metagenomic samples on NCBI SRA globally. Users can select by metagenomic category, view samples on the map, and select samples.

![metagenomes leaflet screenshot](https://github.com/wrf/taxonomy_database/blob/027a5434ea2162068f929d6ba10d92f25b881fa7/images/metagenomes_leaflet_screenshot.jpg)

# [Sea Level Riser](https://github.com/wrf/oceanography_scripts/tree/master/sealevel_rise) #
Version 1.0 of an interactive [ShinyApp](https://rstudio.github.io/leaflet/index.html) map using [leaflet](https://leafletjs.com/) showing sea level rise around the world.

![screenshot of Sea Level Riser v1](https://github.com/wrf/oceanography_scripts/blob/master/sealevel_rise/sea_level_riser_v1_screenshot_over_london.jpg)

# [gene supermatrix orthogroup viewer](https://github.com/wrf/supermatrix) #
Gene orthogroup viewer for supermatrix construction, in understanding occupancy thresholds of the matrix, or finding clusters with high or low occupancy.

![cluster viewer shinyapp screenshot](https://github.com/wrf/supermatrix/blob/ff2d2e381b4aa6c420bf79005ac3541f80c930b3/images/cluster_shinyapp_screenshot.png)

# [Aphrocallistes vastus](https://github.com/PalMuc/Aphrocallistes_vastus_genome/tree/main) DEG explorer #
Explorer for differential gene expression, to change the view, change the p-value threshold (which colors points), select specific genes of interest to analyze the expression levels across samples, select clusters of interest and display the phylogenetic tree of that cluster, and get the raw table items of all of the chosen genes.

![dge viewer shinyapp screenshot](https://github.com/PalMuc/Aphrocallistes_vastus_genome/blob/2ba7512f88a347881c6f6684e7af57d8efd6dd6d/differential_gene_expression/shinyapp/dge_viewer_shinyapp_screenshot_v1.png)

# [Aphrocallistes vastus](https://github.com/PalMuc/Aphrocallistes_vastus_genome/tree/main) ortholog explorer #
Modified version of the supermatrix orthogroup viewer, to display pre-determined phylogenetic trees as well as the domain structure of clusters of interest.

![ortholog cluster browser app screenshot](https://github.com/PalMuc/Aphrocallistes_vastus_genome/blob/2ba7512f88a347881c6f6684e7af57d8efd6dd6d/ortholog_clusters/ortholog_cluster_browser_app_screenshot_01.png)


---


